pkg/azure/plugin.go

```diff
@@ -847,6 +847,211 @@ func (s *azurePluginServer) AttachResource(ctx context.Context, attachResourceRe
+	vnetName := getVnetFromSubnetId(networkInfo.SubnetID)
+	vnet, err := handler.GetVirtualNetwork(ctx, vnetName)
+	if err != nil {
+		utils.Log.Printf("An error occured while getting vnet:%+v", err)
+		return nil, err
+	}
+
+	// Get subnets address spaces
+	localVnetAddressSpaces := []string{}
+	for _, addressSpace := range vnet.Properties.AddressSpace.AddressPrefixes {
+		localVnetAddressSpaces = append(localVnetAddressSpaces, *addressSpace)
+	}
+	if len(localVnetAddressSpaces) == 0 {
+		return nil, fmt.Errorf("unable to get subnet address prefix")
+	}
+
+	missingResources := []*paragliderpb.MissingResource{}
+	visitedIps := map[string]bool{}
+	deletedIps := map[string]bool{}
+
+	requireNAT := false
+	// Get used address spaces of all clouds
+	orchestratorConn, err := grpc.NewClient(s.orchestratorServerAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, fmt.Errorf("unable to establish connection with orchestrator: %w", err)
+	}
+	defer orchestratorConn.Close()
+	orchestratorClient := paragliderpb.NewControllerClient(orchestratorConn)
+	getUsedAddressSpacesResp, err := orchestratorClient.GetUsedAddressSpaces(context.Background(), &emptypb.Empty{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to get used address spaces: %w", err)
+	}
+
+	permitListReq := &paragliderpb.GetPermitListRequest{Namespace: namespace, Resource: resourceID}
+	permitListResp, err := s.GetPermitList(ctx, permitListReq)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, rule := range permitListResp.Rules {
+		peeringCloudInfos, err := utils.GetPermitListRulePeeringCloudInfo(rule, getUsedAddressSpacesResp.AddressSpaceMappings)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get peering cloud infos: %w", err)
+		}
+
+		for i, peeringCloudInfo := range peeringCloudInfos {
+			peerIp := rule.Targets[i]
+
+			// If we're just checking, we don't need to check the same IP twice
+			// if it exists for both inbound and outbound rules
+			// If fixing, both inbound and outbound rules need to be deleted
+			if !shouldFix && visitedIps[peerIp] {
+				continue // Skip if the IP has already been visited and checked
+			}
+
+			if peeringCloudInfo == nil {
+				requireNAT = true
+			} else if peeringCloudInfo.Cloud == utils.AZURE {
+				// Ips don't can't be repeated, so if a deleted Ip is encountered again
+				// in a diff. rule, it's for conn. in the opposite direction and should be deleted
+				if shouldFix && deletedIps[peerIp] {
+					err = handler.DeleteSecurityRule(ctx, *networkInfo.NSG.Name, rule.Name)
+					if err != nil {
+						utils.Log.Printf("An error occured while deleting security rule:%+v", err)
+						return nil, err
+					}
+					continue
+				}
+
+				// Get the URI for the peered resource
+				getIpReq := &paragliderpb.GetUriFromIpRequest{Ip: peerIp, Cloud: utils.AZURE}
```

Comment:
The tags of all resources referenced in a rule are stored in the description. It might be better to query based on the tag instead of the IP?

Reply:
In fact, I don't think there is a way to query by IP in the underlying database structure.

Reply:
It skipped me that we're storing the tags for the permit lists. I'll change this to use the tags instead

Reply:
Also realizing for attached resources, our permit list targets may not have tags stored for them. In that case I'm guessing we can skip them?